### PR TITLE
ARM split - Add i2c_slave support

### PR DESCRIPTION
**Pull request by zvecr**
# While this PR is in draft, the ChibiOS submodule will point to my personal fork

## Current Issue
Transport does not exist

## TL;DR
TODO

<details><summary>CLICK ME FOR INVESTIGATION INFO</summary>

## Transport

Current issues include
* drivers/arm/i2c_serial.[ch] not implemented
  * ChibiOS patch has at least been verified against current QMK fork
      * needs verification against latest release
      * self contained master/slave example might help testing
* split_common/serial.[ch] does not compile for ARM

Requirements
* 1 wire protocol
  * Serial
    * complete implementation required
* 2 wire protocol
  * I2C
    * boards may have wired the I2C bus to both a split slave and device like OLED

Upgrading ChibiOS, maintain patch - could break
Potential to lock in to current version if patch cannot be applied going forward

> Think of the current I2C slave code like a patch file, if the code changes the patch might not apply. Or the surrounding code could change behavior, patch applies fine but fails somehow at runtime.

Requires more in-depth knowledge of ChibiOS than "Contributors/Members" have?

### I2C
Pulling the muon code as a intial example, I2C slave support is reasonably self contained. The current implementation manages one block of memory, and reads/writes at offsets. Each offset is handled as a I2C register. The currently exposed API poses an issue due to ChibiOS not providing a mechninism to know the requested number of bytes. When an I2C callback fires, the amount of data to send is unknown.

Potential Solutions
* Stash offsets -> size, sniff register and map to size
* Refactor api to use a structure which can map data with size
    * Each I2C register can then map 
* Refactor to one block, one register
    * Simple
    * No more offsets, transmit/receive all data in one go
    * More overhead
        * chunking based X bytes

### Serial
Currently coded to only support atmega32u4, and contains timings and various settings which are both compiler. Uses AVR specific interrupts and register configuration. Porting it in its current form would be extremely unlikely. Extracting the existing serial.[ch] into a master/slave under `drivers/avr`, with an exposed API. This can then be implemented for ARM, using the relevant ChibiOS calls.

### Alternatives
While outside of current AVR support, anything that is either simple enough to implement, or bundled as part of ChibiOS could be considered. If it can help bootstrap the project, then its on the table.

* RS232/RS485
  * No ChibiOS support
* SPI
  * Untested patch exists on ChibiOS forum

</details>

## Description
TODO

## Types of Changes

- [x] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by zvecr** [comment]
## Testing notes
Scenarios currently tested
* f103 bluepill split
* f303 split

File: drivers/arm/i2c_defs.h [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+#else
+    // The default PAL alternate modes are used to signal that the pins are used for I2C
+    #ifndef I2C1_SCL_PAL_MODE
+        #define I2C1_SCL_PAL_MODE 4
+    #endif
+    #ifndef I2C1_SDA_PAL_MODE
+        #define I2C1_SDA_PAL_MODE 4
+    #endif
+
+    // The default timing values below configures the I2C clock to 400khz assuming a 72Mhz clock
```
Comment (yanfali) [inline drivers/arm/i2c_defs.h]: this comment may be incorrect now based on configuring 100khz

Comment (yanfali) [reply drivers/arm/i2c_defs.h]: OK, I see how I mis-read this. Do we want to go with 400 for i2cv2? I think bakingpy reported issues with slower MCUs or poor quality cables.

File: keyboards/handwired/splittest/proton_c/readme.md [hunk 2]
```diff
@@ -0,0 +1,11 @@
+# Proton C splittest
+
+To trigger keypress, short together pins *B14* and *B9* (marked on the PCB as *B3* and *B6*).
+
+## Wiring
+- Add switches to both Proton C across B3 and B6 pins
+- Add pull-up resistor to left side between VCC and F6
+- Add pull-down resistors to right side between GND and F6
+- Connect the following pins on both sides together: D0, D1, GND, VCC
```
Comment (yanfali) [inline keyboards/handwired/splittest/proton_c/readme.md]: this is a bit ambiguous, I'm guessing you mean connect the pins on each half straight through?

i.e. A - D0 <---> D0 - B

File: keyboards/handwired/splittest/proton_c/readme.md [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+- Add I2C 4.7kOhm resistors between D0 and VCC, and D1 and VCC
```
Comment (yanfali) [inline keyboards/handwired/splittest/proton_c/readme.md]: Maybe, add pull up resistors on lines D0 and D1; e.g. connect D0 to VCC via a 4.7K resistor

File: keyboards/handwired/splittest/proton_c/readme.md [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+- Wire Di of RGB strip for each half to D3
```
Comment (yanfali) [inline keyboards/handwired/splittest/proton_c/readme.md]: And GND and VCC

**Comment by yanfali** [comment]
Is there a block diagram of the lifecycle of the i2c slave hooks?

File: drivers/arm/i2c_slave.c [hunk 5]
```diff
@@ -0,0 +1,131 @@
+#include "i2c_slave.h"
+#include "quantum.h"
+#include <string.h>
+#include <hal.h>
+#include "hal_i2cslave.h"
+
+#ifndef I2C_SLAVE_TIMEOUT
+  #define I2C_SLAVE_TIMEOUT 500
```
Comment (yanfali) [inline drivers/arm/i2c_slave.c]: millis?

File: drivers/arm/i2c_defs.h (hunk 1 above)
Comment (drashna) [reply drivers/arm/i2c_defs.h]: I think that will need some testing, to be honest. 

But I think for now, 100khz may be the safest solution

Comment (zvecr) [reply drivers/arm/i2c_defs.h]: Ideally, the behaviour should be matched, and ARM I2C should use 400khz, and drop down to 100khz when running split via `quantum/split_common/post_config.h`. When running on a GPIOv1 board, that's easy but GPIOv2 is an issue. The timer configuration is not a single value and hardware dependent.

Now I have a real board, I can at least test the 100/400khz for GPIOv1.

**Comment by zvecr** [comment]
>Is there a block diagram of the lifecycle of the i2c slave hooks?

@yanfali were you after the ChibiOS functionality or the i2c_slave implementation on top?

File: drivers/arm/i2c_slave.c (hunk 5 above)
Comment (drashna) [reply drivers/arm/i2c_slave.c]: yes, should be.

**Comment by yanfali** [comment]
> > Is there a block diagram of the lifecycle of the i2c slave hooks?
> 
> @yanfali were you after the ChibiOS functionality or the i2c_slave implementation on top?

I think you have the most familiarity with this code now, so it would be good for people new to the code to have a shared mental model of the lifecycle.

**Comment by zvecr** [comment]
Given the recent updates break the patch, and we dont really want to support the patch going forward anyway, It would be preferable to target upstream rather than rely on rather invasive changes to the ChibiOS code base. Maybe as its own isolated "doesnt work the same time as master" driver to contrib.

Given the above, I dont see this happening any time soon so closed till upstream has support.

**Comment by dexter93** [comment]
ChibiOS I2C HLD supports slave mode now as of this [patch](https://sourceforge.net/p/chibios/code/16205/)
Perhaps this could be refactored as a base for ARM split support in QMK and the chip specifics can be done in ChibiOS-Contrib.

At a first glance, the past issues with unknown transfer sizes seem to be a non-issue now.
